### PR TITLE
Refactor syscalls

### LINKED ISSUE #1528: Review all syscalls and prepare for release
### File Location(s)

gstd/src/msg/async.rs
etc.
etc.

### Proposal

- [x] Change `gr_send`/`gr_reply` (et al.) syscalls to accept one additional parameter (delay).
- [x] Remove deprecated `gstd::send_and_wait_for_reply` and `gstd::send_bytes_and_wait_for_reply`.
- [x] Rename `gr_wait_no_more` to `gr_wait_up_to`.
- [x] #1208
- [ ] Part of #1065

### PR DESCRIPTION
**Pull request by breathx**
Resolves #1528.
Part of #1065.

Will be finished after #1623

### CODE REVIEW

File: examples/ping-gas/src/lib.rs [hunk 1]
```diff
@@ -6,7 +6,7 @@ use gstd::{debug, msg, prelude::*};
 unsafe extern "C" fn handle() {
     debug!("Hello from ping handle");
 
-    let new_msg = String::from_utf8(msg::load_bytes()).expect("Invalid message");
+    let new_msg = String::from_utf8(msg::load_bytes().unwrap()).expect("Invalid message");
```
Comment (NikVolf) [inline examples/ping-gas/src/lib.rs:9]: no unwraps without an error message please, even in examples

File: gcore/src/prog.rs [hunk 2]
```diff
@@ -18,84 +18,87 @@
...
+            value_ptr: *const u128,
+            delay: u32,
+            message_id_ptr: *mut [u8; 32],
+            program_id_ptr: *mut [u8; 32],
         ) -> SyscallError;
 
+        #[allow(improper_ctypes)]
         pub fn gr_create_program_wgas(
-            code_hash: *const u8,
+            code_id_ptr: *const [u8; 32],
```
Comment (NikVolf) [inline gcore/src/prog.rs:44]: "code_hash" is common term, why introduce "id" here ?

"ptr" suffix is ok
```suggestion
            code_hash_ptr: *const [u8; 32],
```

Comment (breathx) [reply gcore/src/prog.rs:44]: In gear-core we're going with hashes defined inside `MessageId`, `ProgramId` and `CodeId`, so I suggesting renaming of all `CodeHash` entries in order to support similar naming all over the project

File: core-backend/common/src/lib.rs [hunk 3]
```diff
@@ -145,25 +149,25 @@ pub trait RuntimeCtx<E: Ext> {
     fn alloc(&mut self, pages: u32) -> Result<WasmPageNumber, RuntimeCtxError<E::Error>>;
 
     /// Read designated chunk from the memory.
-    fn read_memory(&self, ptr: u32, len: u32) -> Result<Vec<u8>, RuntimeCtxError<E::Error>>;
+    fn read_memory(&self, ptr: i32, len: u32) -> Result<Vec<u8>, RuntimeCtxError<E::Error>>;
```
Comment (breathx) [inline core-backend/common/src/lib.rs:152]: return back to u32

File: core-backend/sandbox/src/funcs.rs [hunk 4]
```diff
@@ -107,21 +73,21 @@ pub enum FuncError<E: Display> {
...
 where
     E: fmt::Display,
 {
-    fn as_core(&self) -> Option<&E> {
-        match self {
-            Self::Core(err) => Some(err),
-            _ => None,
-        }
-    }
+    // fn as_core(&self) -> Option<&E> {
```
Comment (breathx) [inline core-backend/sandbox/src/funcs.rs:85]: remove this

File: core-backend/sandbox/src/funcs.rs [hunk 5]
```diff
@@ -161,485 +127,354 @@ macro_rules! sys_trace {
...
+                Ok(buffer) => {
+                    ctx.memory
+                        .set(buffer_ptr as u32, buffer)
+                        .map_err(|_| MemoryError::OutOfBounds)?;
 
-            if last_idx > msg.len() {
-                return Err(FuncError::ReadWrongRange(at..last_idx, msg.len()));
+                    Ok(0u32)
+                }
+                Err(_err) => Ok(1), // TODO: FIX IT.
```
Comment (breathx) [inline core-backend/sandbox/src/funcs.rs:295]: open issue

File: gcore/src/prog.rs (hunk 2 above)
Comment (breathx) [reply gcore/src/prog.rs:44]: @grishasobol agreed with me.
Do you accept this solution, @NikVolf?

**Review by ark0f — APPROVED** [review]

**Review by grishasobol — APPROVED** [review]

File: gcore/src/prog.rs (hunk 2 above)
Comment (breathx) [reply gcore/src/prog.rs:44]: Decided to leave it as it is atm.
To be changed with no logic changes at any time.

**Pull request merged**